### PR TITLE
Refactor output modes to use enum (#489)

### LINKED ISSUE #489: Refactor output modes to use enum
Currently you can often see in the code the following pattern:
```c
RZ_API void rz_core_task_list(RzCore *core, int mode) {
    if (mode == 'j') {
         ...
    }
}
```
In some cases it looks like that:
```c
RZ_API int rz_core_disasm_pde(RzCore *core, int nb_opcodes, int mode) {
    if (mode == RZ_MODE_JSON) {
        ...
    }
}
```
Sometimes there is no "mode" word, and it looks like that:
```
RZ_API void rz_analysis_xrefs_list(RzAnalysis *analysis, int rad) {
    ...
    if (rad == 'j') {
        ...
    }
```

I recommend to search for `int mode` and `char mode` patterns in the code.

It prevents catching some errors at the compilation time and reduces the readability.
Should be refactored to use enum instead, across the whole Rizin code.

After everything ported, the following hardcoded values should be removed from `librz/include/rz_types.h`:
```c
#define RZ_MODE_PRINT     0x000
#define RZ_MODE_SET       0x002
#define RZ_MODE_SIMPLE    0x004
#define RZ_MODE_JSON      0x008
#define RZ_MODE_SIMPLEST  0x020
```

See, for example:
```
librz/util/range.c
314:int rz_range_list(RRange *rgs, int rad) {

librz/io/io_cache.c
93:RZ_API bool rz_io_cache_list(RzIO *io, int rad) {

librz/main/rz-bin.c
63:static RzOutputMode rad2outputmode(int rad) {
471:static int rabin_do_operation(RzBin *bin, const char *op, int rad, const char *output, const char *file) {
622:static void __listPlugins(RzBin *bin, const char *plugin_name, PJ *pj, int rad) {
656:    int rad = 0;

librz/core/core_private.h
123:RZ_IPI bool rz_core_debug_reg_list(RzCore *core, int type, int size, bool skip_covered, PJ *pj, int rad, const char *use_color);

librz/core/cdebug.c
300:RZ_IPI bool rz_core_debug_reg_list(RzCore *core, int type, int size, bool skip_covered, PJ *pj, int rad, const char *use_color) {

librz/core/cmd/cmd_flag.c
734:    int rad;
759:static void print_function_labels_for(RzAnalysisFunction *fcn, int rad, PJ *pj) {
772:static void print_function_labels(RzAnalysis *analysis, RzAnalysisFunction *fcn, int rad) {

librz/core/canalysis.c
4157:static bool found_xref(RzCore *core, ut64 at, ut64 xref_to, RzAnalysisXRefType type, PJ *pj, int rad, int cfg_debug, bool cfg_analysis_strings) {
4225:RZ_API int rz_core_analysis_search_xrefs(RzCore *core, ut64 from, ut64 to, PJ *pj, int rad) {

librz/core/cmd/cmd_analysis.c
7005:   int rad;

librz/debug/ddesc.c
71:RZ_API int rz_debug_desc_list(RzDebug *dbg, int rad) {

librz/flag/flag.c
374:RZ_API void rz_flag_list(RzFlag *f, int rad, const char *pfx) {

librz/debug/p/bfvm.c
273:RZ_API void bfvm_show_regs(BfvmCPU *c, int rad) {
289:RZ_API void bfvm_maps(BfvmCPU *c, int rad) {

librz/cons/pal.c
526:RZ_API void rz_cons_pal_list(int rad, const char *arg) {

librz/include/rz_flag.h
97:RZ_API void rz_flag_list(RzFlag *f, int rad, const char *pfx);

librz/include/rz_debug.h
518:RZ_API int rz_debug_desc_list(RzDebug *dbg, int rad);

librz/include/rz_cons.h
953:RZ_API void rz_cons_pal_list(int rad, const char *arg);

librz/include/rz_analysis.h
55:     int rad;

librz/include/rz_io.h
394:RZ_API bool rz_io_cache_list(RzIO *io, int rad);

librz/include/rz_core.h
670:RZ_API int rz_core_analysis_search_xrefs(RzCore *core, ut64 from, ut64 to, PJ *pj, int rad);

librz/include/rz_util/rz_range.h
35:RZ_API int rz_range_list(RRange *rgs, int rad);
```

### PR DESCRIPTION
**Pull request by valdaarhun**
**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Refactored output modes to use enum. `int mode` and `char mode` patterns in the code have been replaced with `RzOutputMode` enum type.

*Refactoring is still in progress.*

**Test plan**

...

**Closing issues**

closes #489

### CODE REVIEW

File: librz/analysis/meta.c [hunk 1]
```diff
@@ -237,8 +237,8 @@ RZ_API const char *rz_meta_type_to_string(int type) {
 	return "# unknown meta # ";
 }
 
-RZ_API void rz_meta_print(RzAnalysis *a, RzAnalysisMetaItem *d, ut64 start, ut64 size, int rad, PJ *pj, bool show_full) {
-	rz_return_if_fail(!(rad == 'j' && !pj)); // rad == 'j' => pj != NULL
+RZ_API void rz_meta_print(RzAnalysis *a, RzAnalysisMetaItem *d, ut64 start, ut64 size, RzOutputMode rad, PJ *pj, bool show_full) {
```
Comment (XVilka) [inline librz/analysis/meta.c]: I suggest to rename variable too `rad` -> `mode`. And everywhere else you meet unclear name for the mode.

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/analysis/meta.c (hunk 1 above)
Comment (valdaarhun) [reply librz/analysis/meta.c]: Ok, will do.

File: librz/analysis/meta.c [hunk 2]
```diff
@@ -339,7 +339,7 @@ RZ_API void rz_meta_print(RzAnalysis *a, RzAnalysisMetaItem *d, ut64 start, ut64
 			break;
 		case 0:
 		case 1:
```
Comment (wargio) [inline librz/analysis/meta.c:341]: i would remove 0 and 1.

File: librz/analysis/meta.c [hunk 3]
```diff
@@ -339,7 +339,7 @@ RZ_API void rz_meta_print(RzAnalysis *a, RzAnalysisMetaItem *d, ut64 start, ut64
...
-		case '*':
+		case RZ_OUTPUT_MODE_RIZIN:
```
Comment (wargio) [inline librz/analysis/meta.c:342]: no need to define this if it goes into the default switch case.

File: librz/bin/pdb/pdb.c [hunk 4]
```diff
@@ -1317,8 +1317,8 @@ static void print_types(const RzPdb *pdb, PJ *pj, const int mode) {
 	}
 	switch (mode) {
 	case 'd': print_types_regular(pdb, tpi_stream->types); return;
-	case 'j': print_types_json(pdb, pj, tpi_stream->types); return;
-	case 'r': print_types_format(pdb, tpi_stream->types); return;
+	case RZ_OUTPUT_MODE_JSON: print_types_json(pdb, pj, tpi_stream->types); return;
+	case RZ_OUTPUT_MODE_RIZIN: print_types_format(pdb, tpi_stream->types); return;
```
Comment (wargio) [inline librz/bin/pdb/pdb.c:1321]: i'm not sure about this switch. what currently does the `'d`' command?

File: librz/core/canalysis.c [hunk 5]
```diff
@@ -3789,9 +3789,9 @@ static int fcn_print_legacy(RzCore *core, RzAnalysisFunction *fcn) {
 		var_count += rz_analysis_var_count(core->analysis, fcn, 'r', 0);
 
 		rz_cons_printf("\nlocals: %d\nargs: %d\n", var_count, args_count);
-		rz_analysis_var_list_show(core->analysis, fcn, 'b', 0, NULL);
-		rz_analysis_var_list_show(core->analysis, fcn, 's', 0, NULL);
-		rz_analysis_var_list_show(core->analysis, fcn, 'r', 0, NULL);
+		rz_analysis_var_list_show(core->analysis, fcn, 'b', RZ_OUTPUT_MODE_QUIET, NULL);
+		rz_analysis_var_list_show(core->analysis, fcn, 's', RZ_OUTPUT_MODE_QUIET, NULL);
+		rz_analysis_var_list_show(core->analysis, fcn, 'r', RZ_OUTPUT_MODE_QUIET, NULL);
```
Comment (wargio) [inline librz/core/canalysis.c:3794]: i think this should be `RZ_OUTPUT_MODE_STANDARD`

File: librz/core/canalysis.c [hunk 6]
```diff
@@ -4464,11 +4464,11 @@ RZ_API int rz_core_analysis_search(RzCore *core, ut64 from, ut64 to, ut64 ref, i
 						op.size = 1;
 					}
 					break;
-				case 'r':
+				case RZ_OUTPUT_MODE_RIZIN:
```
Comment (wargio) [inline librz/core/canalysis.c:4467]: i'm not sure this change is correct. that 'r' is a command, not a mode.

File: librz/core/cmd_api.c [hunk 7]
```diff
@@ -260,7 +260,7 @@ static RzOutputMode suffix2mode(const char *suffix) {
 			return argv_modes[i].mode;
 		}
 	}
-	return 0;
+	return RZ_OUTPUT_MODE_QUIET;
```
Comment (wargio) [inline librz/core/cmd_api.c:263]: not sure about this change either.

File: librz/core/cmd_debug.c [hunk 8]
```diff
@@ -1300,8 +1300,8 @@ static void cmd_debug_modules(RzCore *core, int mode) { // "dmm"
 			pj_end(pj);
 		} break;
 		case ':':
-		case '*':
-			if (mode == '*' || (mode == ':' && addr >= map->addr && addr < map->addr_end)) {
+		case 'RZ_OUTPUT_MODE_RIZIN':
```
Comment (wargio) [inline librz/core/cmd_debug.c:1303]: this change doesn't make sense. the command requires to be splitted.

File: librz/analysis/meta.c (hunk 1 above)
Comment (valdaarhun) [reply librz/analysis/meta.c]: I have changed `rad` to `mode` wherever output modes are involved.

File: librz/analysis/meta.c (hunk 3 above)
Comment (valdaarhun) [reply librz/analysis/meta.c:342]: Ok, so I will remove all the three cases.

File: librz/analysis/meta.c [hunk 9]
```diff
@@ -237,8 +237,8 @@ RZ_API const char *rz_meta_type_to_string(int type) {
 	return "# unknown meta # ";
 }
 
-RZ_API void rz_meta_print(RzAnalysis *a, RzAnalysisMetaItem *d, ut64 start, ut64 size, int rad, PJ *pj, bool show_full) {
-	rz_return_if_fail(!(rad == 'j' && !pj)); // rad == 'j' => pj != NULL
+RZ_API void rz_meta_print(RzAnalysis *a, RzAnalysisMetaItem *d, ut64 start, ut64 size, RzOutputMode mode, PJ *pj, bool show_full) {
+	rz_return_if_fail(!(mode == RZ_OUTPUT_MODE_JSON && !pj)); // mode == 'j' => pj != NULL
```
Comment (XVilka) [inline librz/analysis/meta.c:241]: Please also update the commentary or just remove it.

File: librz/analysis/meta.c [hunk 10]
```diff
@@ -348,7 +348,7 @@ RZ_API void rz_meta_print(RzAnalysis *a, RzAnalysisMetaItem *d, ut64 start, ut64
 				if (!s) {
 					s = strdup(pstr);
 				}
-				if (rad) {
+				if (mode) {
```
Comment (XVilka) [inline librz/analysis/meta.c:351]: I don't think this should be `if (mode)`, code doesn't make sense in JSON mode, for example. Lets just use some specific mode here.

File: librz/analysis/meta.c [hunk 11]
```diff
@@ -370,7 +370,7 @@ RZ_API void rz_meta_print(RzAnalysis *a, RzAnalysisMetaItem *d, ut64 start, ut64
 				free(s);
 			} break;
 			case RZ_META_TYPE_STRING:
-				if (rad) {
+				if (mode) {
```
Comment (XVilka) [inline librz/analysis/meta.c:373]: Same here and below.

File: librz/bp/bp.c [hunk 12]
```diff
@@ -306,7 +306,7 @@ RZ_API int rz_bp_list(RzBreakpoint *bp, int rad) {
 			pj_ks(pj, "data", rz_str_get(b->data));
 			pj_ks(pj, "cond", rz_str_get(b->cond));
 			pj_end(pj);
-		} else if (rad) {
+		} else if (mode) {
```
Comment (XVilka) [inline librz/bp/bp.c:309]: Same here, better to use a switch instead of `if`/`else` and use specific modes.

File: librz/core/cmd_flag.c [hunk 13]
```diff
@@ -1216,7 +1216,35 @@ RZ_IPI int rz_cmd_flag(void *data, const char *input) {
 			}
 			break;
 		case 's':
-			flag_space_stack_list(core->flags, input[2]);
+			RzOutputMode mode;
```
Comment (XVilka) [inline librz/core/cmd_flag.c:1219]: You could extract this in a separate function

File: librz/core/cmd_flag.c [hunk 14]
```diff
@@ -1456,7 +1484,35 @@ RZ_IPI int rz_cmd_flag(void *data, const char *input) {
 				free(s);
 			}
 		} else {
-			rz_flag_list(core->flags, *input, input[0] ? input + 1 : "");
+			RzOutputMode mode;
```
Comment (XVilka) [inline librz/core/cmd_flag.c:1487]: Same here, will also avoid duplication.

File: librz/core/cmd_meta.c [hunk 15]
```diff
@@ -664,18 +664,73 @@ static int cmd_meta_others(RzCore *core, const char *input) {
 		}
 	} break;
 	case '.': // "Cf.", "Cd.", ...
+		RzOutputMode mode = RZ_OUTPUT_MODE_STANDARD;
 		if (input[2] == '.') { // "Cs.."
 			ut64 size;
 			RzAnalysisMetaItem *mi = rz_meta_get_at(core->analysis, addr, type, &size);
 			if (mi) {
-				rz_meta_print(core->analysis, mi, addr, size, input[3], NULL, false);
+				switch(input[3]){
```
Comment (XVilka) [inline librz/core/cmd_meta.c:672]: And here

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cmd_debug.c (hunk 8 above)
Comment (valdaarhun) [reply librz/core/cmd_debug.c:1303]: @wargio I am sorry but I haven't understood what you mean by "command should be split". Do you mean I should have a separate case for `mode == ':'`? Also, I noticed that I have accidentally put `RZ_OUTPUT_MODE_RIZIN` in single quotes. I'll change that as well.

File: librz/analysis/meta.c (hunk 10 above)
Comment (valdaarhun) [reply librz/analysis/meta.c:351]: @XVilka Do you mean I should change it to something more specific like `if (mode == RZ_OUTPUT_MODE_STANDARD`)?

File: librz/core/cmd_flag.c (hunk 13 above)
Comment (valdaarhun) [reply librz/core/cmd_flag.c:1219]: @XVilka I just wanted to confirm one thing. Should I create an `RZ_API` defined function in this file that I can then call in every other file as well?

**Comment by wargio** [comment]
ok, here is what i would do, because this PR is quite messy and big even to comment.

1. I would keep this branch and switch to `dev`
2. i would then create a new branch from dev
3. i would edit only 1 function at time, fix any incompatibility with the old code, rebuild and run the tests.
4. if the tests does not fail, then commit these changes and push them.
5. repeat 3 and 4 till you have converted all the functions and made a single commit for each of them.

you can open a new PR and reference that one to this.

**Comment by wargio** [comment]
if you prefer, we can go function by function, so we can guide you towards the resolution of the issue.

**Comment by XVilka** [comment]
I agree about step-by-step approach, but instead of one function would do one/two files per PR. This way export from this PR is easy - just use `git diff filename` or similar command.

**Comment by valdaarhun** [comment]
@wargio @XVilka Yeah, I think I'll do that. It'll be more organised that way.

**Comment by valdaarhun** [comment]
@wargio @XVilka I created a new branch and refactored one file. I built it on my local computer and didn't get any compilation error, so I decided to push it in a new PR (#950). However, I have got three failing tests and I don't understand the reason behind them.

**Comment by wargio** [comment]
i'll close this to discuss on the other PR.

**Pull request closed without merging**